tasks/libs/common/gitlab.py

```diff
@@ -56,23 +56,41 @@ def create_pipeline(self, project_name, ref, variables=None):
         data = json.dumps({"ref": ref, "variables": [{"key": k, "value": v} for (k, v) in variables.items()],})
         return self.make_request(path, headers=headers, data=data, json=True)
 
-    def pipelines_for_ref(self, project_name, ref, per_page=100):
+    def all_pipelines_for_ref(self, project_name, ref, sha=None):
         """
-        Gets all pipelines for a given reference
+        Gets all pipelines for a given reference (+ optionally git sha).
+        """
+        pipelines = []
+        page = 1
+
+        # Go through all pages
+        results = self.pipelines_for_ref(project_name, ref, sha=sha, page=page)
+        while results:
+            pipelines.extend(results)
+            page += 1
+            results = self.pipelines_for_ref(project_name, ref, sha=sha, page=page)
+
+        return pipelines
+
+    def pipelines_for_ref(self, project_name, ref, sha=None, page=1, per_page=100):
+        """
+        Gets one page of pipelines for a given reference (+ optionally git sha).
         """
         from urllib.parse import quote
 
-        path = "/projects/{}/pipelines?ref={}&per_page={}".format(
-            quote(project_name, safe=""), quote(ref, safe=""), per_page,
+        path = "/projects/{}/pipelines?ref={}&per_page={}&page={}".format(
+            quote(project_name, safe=""), quote(ref, safe=""), per_page, page
         )
+        if sha:
+            path = "{}&sha={}".format(path, sha)
```

Comment:
Maybe not on this PR but we could/should rely on the requests `param` argument to avoid handling URL arguments manually